File: packages/synesis-context-trust/src/scanner.ts
/**
 * Unified pattern scanner — TS port of guardrails_core/scanner.py.
 *
 * Three tiers of compiled patterns:
 *   Tier 1 (core): instruction override, jailbreak, template injection
 *   Tier 2 (web/indirect): encoded payloads, link injection, hidden text, prompt leaking
 *   Tier 3 (output): signs the model complied with an injection
 *
 * Keep pattern lists in sync with the Python implementation.
 */

import { normalizeForScan, detectBase64Payloads } from "./normalizer.js";

// ---------------------------------------------------------------------------
// Tier 1: Core injection patterns
// ---------------------------------------------------------------------------
const CORE_PATTERNS: RegExp[] = [
  /ignore\s+(?:all\s+)?(?:previous|prior|above)\s+instructions?/i,
  /disregard\s+(?:all\s+)?(?:previous|prior|above)/i,
  /forget\s+(?:everything|all)\s+(?:you\s+)?(?:were\s+)?told/i,
  /new\s+instructions?\s*:/i,
  /override\s+(?:your\s+)?(?:instructions?|prompt)/i,
  /you\s+are\s+now\s+(?:a|an)\s/i,
  /pretend\s+you\s+are/i,
  /act\s+as\s+if\s+you/i,
  /(?:^|\n)\s*system\s*:\s*(?:ignore|disregard|forget|override|follow\s+these\s+instructions|you\s+are\s+now|pretend|act\s+as)/i,
  /<\|im_start\|>\s*system/i,
  /###\s*human\s*:/i,
  /\[INST\]\s*/i,
  /<\/?s(?:ystem)?>/i,
  /ignore\s+the\s+above/i,
  /ignore\s+above\b/i,
  /follow\s+these\s+instructions?\s+instead/i,
  /output\s+(?:only|just)\s+the\s+following/i,
  /print\s+(?:exactly|only)\s+this\s*:/i,
  /(?:DAN|developer)\s+mode\s+(?:enabled|activated|on)/i,
  /(?:do\s+anything\s+now|unlimited\s+mode)/i,
];

// ---------------------------------------------------------------------------
// Tier 2: Extended patterns (web, indirect injection, obfuscation)
// ---------------------------------------------------------------------------
const WEB_PATTERNS: RegExp[] = [
  /base64[:\s]+[A-Za-z0-9+/=]{20,}/i,
  /\[[^\]\r\n]{0,2048}\]\(\s*javascript\s*:/i,
  /<a\b[^>\r\n]{0,2048}\bhref\s*=\s*["']?\s*javascript:/i,
  /[\u200b\u200c\u200d\u2060\ufeff]{3,}/,
  /data:text\/html[;,]/i,
  /(?:reveal|show|print|repeat|echo)\s+(?:your\s+)?(?:system\s+)?(?:prompt|instructions)/i,
  /what\s+(?:are|is)\s+your\s+(?:system\s+)?(?:prompt|instructions)/i,
  /from\s+now\s+on\s+(?:you\s+)?(?:are|will|must|should)\b/i,
  /(?:assistant|ai|model)\s*:\s*(?:sure|okay|yes|I will)/i,
  /<!--\s*(?:system|instruction|prompt)/i,
];

// ---------------------------------------------------------------------------
// Tier 3: Output compliance indicators
// ---------------------------------------------------------------------------
const OUTPUT_PATTERNS: RegExp[] = [
  /(?:^|\n)\s*system\s*:\s*(?:you\s+are|ignore|disregard|forget|override|follow\s+these\s+instructions)/i,
  /(?:my|the)\s+system\s+prompt\s+(?:is|says|reads)/i,
  /(?:here\s+(?:is|are)\s+)?my\s+(?:original\s+)?instructions?:/i,
  /I\s+(?:will|can|shall)\s+now\s+(?:act|behave|operate)\s+as/i,
  /(?:DAN|developer)\s+mode\s+(?:enabled|activated)/i,
  /<\|im_start\|>/i,
];

const CODE_NOISE_PATTERNS: RegExp[] = [
  /^(show|print|repeat|echo)\s+(?:system\s+)?prompt$/i,
  /^(show|print|repeat|echo)\s+instructions?$/i,
];

const CORE_REDACTION_PATTERNS: RegExp[] = [
  /ignore\s+(?:all\s+)?(?:previous|prior|above)\s+instructions?/gi,
  /disregard\s+(?:all\s+)?(?:previous|prior|above)/gi,
  /forget\s+(?:everything|all)\s+(?:you\s+)?(?:were\s+)?told/gi,
  /new\s+instructions?\s*:/gi,
  /override\s+(?:your\s+)?(?:instructions?|prompt)/gi,
  /you\s+are\s+now\s+(?:a|an)\s/gi,
  /pretend\s+you\s+are/gi,
  /act\s+as\s+if\s+you/gi,
  /(?:^|\n)\s*system\s*:\s*(?:ignore|disregard|forget|override|follow\s+these\s+instructions|you\s+are\s+now|pretend|act\s+as)/gi,
  /<\|im_start\|>\s*system/gi,
  /###\s*human\s*:/gi,
  /\[INST\]\s*/gi,
  /<\/?s(?:ystem)?>/gi,
  /ignore\s+the\s+above/gi,
  /ignore\s+above\b/gi,
  /follow\s+these\s+instructions?\s+instead/gi,
  /output\s+(?:only|just)\s+the\s+following/gi,
  /print\s+(?:exactly|only)\s+this\s*:/gi,
  /(?:DAN|developer)\s+mode\s+(?:enabled|activated|on)/gi,
  /(?:do\s+anything\s+now|unlimited\s+mode)/gi,
];

const WEB_REDACTION_PATTERNS: RegExp[] = [
  /base64[:\s]+[A-Za-z0-9+/=]{20,}/gi,
  /\[[^\]\r\n]{0,2048}\]\(\s*javascript\s*:/gi,
  /<a\b[^>\r\n]{0,2048}\bhref\s*=\s*["']?\s*javascript:/gi,
  /[\u200b\u200c\u200d\u2060\ufeff]{3,}/g,
  /data:text\/html[;,]/gi,
  /(?:reveal|show|print|repeat|echo)\s+(?:your\s+)?(?:system\s+)?(?:prompt|instructions)/gi,
  /what\s+(?:are|is)\s+your\s+(?:system\s+)?(?:prompt|instructions)/gi,
  /from\s+now\s+on\s+(?:you\s+)?(?:are|will|must|should)\b/gi,
  /(?:assistant|ai|model)\s*:\s*(?:sure|okay|yes|I will)/gi,
  /<!--\s*(?:system|instruction|prompt)/gi,
];

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type EventType =
  | "system_override_attempt"
  | "jailbreak_roleplay"
  | "context_confusion_attack"
  | "code_exec_risk"
  | "prompt_leakage_attempt"
  | "unknown";

export interface ScanResult {
  detected: boolean;
  patterns_found: string[];
  source: string;
  excerpt: string;
  tier: string;
  confidence: number;
  event_type: EventType;
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

function classifyPatterns(patternsFound: string[]): EventType {
  const joined = patternsFound.join(" ").toLowerCase();
  if (["ignore", "disregard", "override", "new instructions"].some((w) => joined.includes(w)))
    return "system_override_attempt";
  if (["dan", "pretend", "act as", "you are now", "unlimited"].some((w) => joined.includes(w)))
    return "jailbreak_roleplay";
  if (["reveal", "show", "print", "repeat", "echo", "prompt", "instructions"].some((w) => joined.includes(w)))
    return "prompt_leakage_attempt";
  if (["base64", "javascript", "data:text"].some((w) => joined.includes(w)))
    return "code_exec_risk";
  if (["im_start", "inst", "system:"].some((w) => joined.includes(w)))
    return "context_confusion_attack";
  return "unknown";
}

function runPatterns(text: string, patterns: RegExp[], maxChars = 32_000): string[] {
  const found: string[] = [];
  const chunk = text.slice(0, maxChars);
  for (const pat of patterns) {
    const m = pat.exec(chunk);
    if (m) found.push(m[0].slice(0, 80));
  }
  return found;
}

function excerptAround(text: string, patterns: RegExp[], maxChars = 32_000): string {
  const chunk = text.slice(0, maxChars);
  for (const pat of patterns) {
    const m = pat.exec(chunk);
    if (m) {
      const start = Math.max(0, m.index - 50);
      const end = Math.min(chunk.length, m.index + m[0].length + 50);
      return chunk.slice(start, end).replace(/\n/g, " ");
    }
  }
  return "";
}

function isCodeLike(text: string): boolean {
  const sample = text.slice(0, 4000);
  if (!sample.trim()) return false;
  const lines = sample.split(/\r?\n/).slice(0, 80);
  let score = 0;
  for (const raw of lines) {
    const line = raw.trim();
    if (!line) continue;
    if (/^(\/\/|#|\/\*|\*|--)\s*/.test(line)) score += 1;
    if (/[{}()[\];]/.test(line)) score += 1;
    if (/(^|\s)(func|class|def|const|let|var|return|if|for|while|import|export|package)\b/.test(line)) score += 1;
    if (/[:=]-?=?|=>/.test(line)) score += 1;
    if (/\b[A-Za-z_]\w*\s*\(/.test(line)) score += 1;
  }
  return score >= 4;
}

function filterCodeNoise(matches: string[]): string[] {
  return matches.filter((m) => !CODE_NOISE_PATTERNS.some((pat) => pat.test(m.trim())));
}

// ---------------------------------------------------------------------------
// Public scan functions
// ---------------------------------------------------------------------------

export function scanText(text: string, source = "unknown", maxScanChars = 32_000): ScanResult {
  if (!text) return { detected: false, patterns_found: [], source, excerpt: "", tier: "core", confidence: 0, event_type: "unknown" };
  const found = runPatterns(text, CORE_PATTERNS, maxScanChars);
  const excerpt = found.length > 0 ? excerptAround(text, CORE_PATTERNS, maxScanChars) : "";
  const eventType = found.length > 0 ? classifyPatterns(found) : "unknown";
  const confidence = found.length > 0 ? Math.min(0.5 + 0.15 * found.length, 1.0) : 0;
  return { detected: found.length > 0, patterns_found: found, source, excerpt, tier: "core", confidence, event_type: eventType };
}

export function scanWebContent(text: string, source = "web", maxScanChars = 32_000): ScanResult {
  if (!text) return { detected: false, patterns_found: [], source, excerpt: "", tier: "web", confidence: 0, event_type: "unknown" };

  const rawWebFound = runPatterns(text.slice(0, maxScanChars), WEB_PATTERNS, maxScanChars);
  const normalized = normalizeForScan(text);
  const coreFound = runPatterns(normalized, CORE_PATTERNS, maxScanChars);
  const webFound = runPatterns(normalized, WEB_PATTERNS, maxScanChars);

  const seen = new Set(webFound);
  for (const f of rawWebFound) {
    if (!seen.has(f)) {
      webFound.push(f);
      seen.add(f);
    }
  }

  const b64Found = detectBase64Payloads(text, CORE_PATTERNS, maxScanChars);
  let allFound = [...coreFound, ...webFound, ...b64Found];
  if (isCodeLike(text)) {
    allFound = filterCodeNoise(allFound);
  }
  const excerpt = allFound.length > 0 ? excerptAround(normalized, [...CORE_PATTERNS, ...WEB_PATTERNS], maxScanChars) : "";
  const eventType = allFound.length > 0 ? classifyPatterns(allFound) : "unknown";
  const confidence = allFound.length > 0 ? Math.min(0.5 + 0.12 * allFound.length, 1.0) : 0;
  return { detected: allFound.length > 0, patterns_found: allFound, source, excerpt, tier: "web", confidence, event_type: eventType };
}

export function scanModelOutput(text: string, source = "model_output"): ScanResult {
  if (!text) return { detected: false, patterns_found: [], source, excerpt: "", tier: "output", confidence: 0, event_type: "unknown" };
  const found = runPatterns(text, OUTPUT_PATTERNS, 16_000);
  const excerpt = found.length > 0 ? excerptAround(text, OUTPUT_PATTERNS, 16_000) : "";
  const confidence = found.length > 0 ? Math.min(0.6 + 0.15 * found.length, 1.0) : 0;
  return { detected: found.length > 0, patterns_found: found, source, excerpt, tier: "output", confidence, event_type: found.length > 0 ? "prompt_leakage_attempt" : "unknown" };
}

export function redactPatterns(text: string, includeWeb = false): string {
  const patterns = includeWeb ? [...CORE_REDACTION_PATTERNS, ...WEB_REDACTION_PATTERNS] : CORE_REDACTION_PATTERNS;
  let result = text;
  for (const pat of patterns) {
    result = result.replace(pat, "[REDACTED]");
  }
  return result;
}

/**
 * Scan user input and recent conversation history for injection.
 * Returns [detected, scanDetails].
 */
export function scanUserInput(
  userContent: string,
  conversationHistory: string[],
): [boolean, { detected: boolean; patterns_found: string[]; source: string }] {
  const results: ScanResult[] = [];

  if (userContent) {
    const r = scanText(userContent, "user_message");
    results.push(r);
  }

  for (let i = 0; i < conversationHistory.length; i++) {
    if (conversationHistory[i]) {
      const r = scanText(conversationHistory[i], `history_${i}`);
      if (r.detected) results.push(r);
    }
  }

  const detected = results.some((r) => r.detected);
  const allPatterns = results.flatMap((r) => r.patterns_found);
  return [detected, { detected, patterns_found: allPatterns, source: results[0]?.source ?? "user_message" }];
}

File: base/security/guardrails_core/scanner.py
"""Unified pattern scanner — merged from Planner and Yarn injection scanners.

Three tiers of compiled patterns:
  Tier 1 (core): instruction override, jailbreak, template injection
  Tier 2 (web/indirect): encoded payloads, link injection, hidden text, prompt leaking
  Tier 3 (output): signs the model complied with an injection
"""

from __future__ import annotations

import re
import time
from typing import Any

from .normalizer import detect_base64_payloads, normalize_for_scan
from .schemas import EventType, ScanResult

# ---------------------------------------------------------------------------
# Tier 1: Core injection patterns
# ---------------------------------------------------------------------------
CORE_PATTERNS: list[re.Pattern[str]] = [
    re.compile(r"ignore\s+(?:all\s+)?(?:previous|prior|above)\s+instructions?", re.IGNORECASE),
    re.compile(r"disregard\s+(?:all\s+)?(?:previous|prior|above)", re.IGNORECASE),
    re.compile(r"forget\s+(?:everything|all)\s+(?:you\s+)?(?:were\s+)?told", re.IGNORECASE),
    re.compile(r"new\s+instructions?\s*:", re.IGNORECASE),
    re.compile(r"override\s+(?:your\s+)?(?:instructions?|prompt)", re.IGNORECASE),
    re.compile(r"you\s+are\s+now\s+(?:a|an)\s", re.IGNORECASE),
    re.compile(r"pretend\s+you\s+are", re.IGNORECASE),
    re.compile(r"act\s+as\s+if\s+you", re.IGNORECASE),
    re.compile(
        r"(?:^|\n)\s*system\s*:\s*(?:ignore|disregard|forget|override|follow\s+these\s+instructions|you\s+are\s+now|pretend|act\s+as)",
        re.IGNORECASE,
    ),
    re.compile(r"<\|im_start\|>\s*system", re.IGNORECASE),
    re.compile(r"###\s*human\s*:", re.IGNORECASE),
    re.compile(r"\[INST\]\s*", re.IGNORECASE),
    re.compile(r"<\/?s(?:ystem)?>", re.IGNORECASE),
    re.compile(r"ignore\s+the\s+above", re.IGNORECASE),
    re.compile(r"ignore\s+above\b", re.IGNORECASE),
    re.compile(r"follow\s+these\s+instructions?\s+instead", re.IGNORECASE),
    re.compile(r"output\s+(?:only|just)\s+the\s+following", re.IGNORECASE),
    re.compile(r"print\s+(?:exactly|only)\s+this\s*:", re.IGNORECASE),
    re.compile(r"(?:DAN|developer)\s+mode\s+(?:enabled|activated|on)", re.IGNORECASE),
    re.compile(r"(?:do\s+anything\s+now|unlimited\s+mode)", re.IGNORECASE),
    re.compile(
        r"\b(?:send|post|upload|exfiltrate|leak|forward|transmit)\b[\s\S]{0,120}\b(?:secrets?|tokens?|api[_ -]?keys?|passwords?|credentials?|\.env|private\s+keys?)\b[\s\S]{0,120}\b(?:https?:\/\/|webhook|external|remote|attacker|requestbin|pastebin)\b",
        re.IGNORECASE,
    ),
    re.compile(
        r"\b(?:read|open|cat|print|dump|show)\b[\s\S]{0,80}\b(?:\.env\b|~\/\.ssh|id_rsa|id_ed25519|private[_ -]?key|aws[_ -]?credentials|kubeconfig|\/etc\/passwd)\b",
        re.IGNORECASE,
    ),
]

# ---------------------------------------------------------------------------
# Tier 2: Extended patterns (web, indirect injection, obfuscation)
# ---------------------------------------------------------------------------
WEB_PATTERNS: list[re.Pattern[str]] = [
    re.compile(r"base64[:\s]+[A-Za-z0-9+/=]{20,}", re.IGNORECASE),
    re.compile(r"\[[^\]\r\n]{0,2048}\]\(\s*javascript\s*:", re.IGNORECASE),
    re.compile(r"<a\b[^>\r\n]{0,2048}\bhref\s*=\s*[\"']?\s*javascript:", re.IGNORECASE),
    re.compile(r"[\u200b\u200c\u200d\u2060\ufeff]{3,}"),
    re.compile(r"data:text/html[;,]", re.IGNORECASE),
    re.compile(r"(?:reveal|show|print|repeat|echo)\s+(?:your\s+)?(?:system\s+)?(?:prompt|instructions)", re.IGNORECASE),
    re.compile(r"what\s+(?:are|is)\s+your\s+(?:system\s+)?(?:prompt|instructions)", re.IGNORECASE),
    re.compile(r"from\s+now\s+on\s+(?:you\s+)?(?:are|will|must|should)\b", re.IGNORECASE),
    re.compile(r"(?:assistant|ai|model)\s*:\s*(?:sure|okay|yes|I will)", re.IGNORECASE),
    re.compile(r"<!--\s*(?:system|instruction|prompt)", re.IGNORECASE),
    re.compile(
        r"(?:display\s*:\s*none|visibility\s*:\s*hidden|font-size\s*:\s*0)[^<\n]{0,240}(?:ignore|system|instruction|prompt)",
        re.IGNORECASE,
    ),
    re.compile(
        r"!\[[^\]\r\n]{0,2048}\]\(\s*https?:\/\/[^\s)]{0,2048}(?:token|secret|api[_-]?key|password|env)=", re.IGNORECASE
    ),
]

# ---------------------------------------------------------------------------
# Tier 3: Output compliance indicators
# ---------------------------------------------------------------------------
OUTPUT_PATTERNS: list[re.Pattern[str]] = [
    re.compile(
        r"(?:^|\n)\s*system\s*:\s*(?:you\s+are|ignore|disregard|forget|override|follow\s+these\s+instructions)",
        re.IGNORECASE,
    ),
    re.compile(r"(?:my|the)\s+system\s+prompt\s+(?:is|says|reads)", re.IGNORECASE),
    re.compile(r"(?:here\s+(?:is|are)\s+)?my\s+(?:original\s+)?instructions?:", re.IGNORECASE),
    re.compile(r"I\s+(?:will|can|shall)\s+now\s+(?:act|behave|operate)\s+as", re.IGNORECASE),
    re.compile(r"(?:DAN|developer)\s+mode\s+(?:enabled|activated)", re.IGNORECASE),
    re.compile(r"<\|im_start\|>", re.IGNORECASE),
    re.compile(
        r"\b(?:OPENAI_API_KEY|ANTHROPIC_API_KEY|AWS_SECRET_ACCESS_KEY|GITHUB_TOKEN|NPM_TOKEN|DATABASE_URL)\s*[:=]\s*['\"]?[A-Za-z0-9_./+=:-]{12,}",
        re.IGNORECASE,
    ),
    re.compile(r"-----BEGIN\s+(?:RSA\s+|OPENSSH\s+|EC\s+|DSA\s+)?PRIVATE\s+KEY-----", re.IGNORECASE),
    re.compile(r"\b(?:sk-[A-Za-z0-9]{20,}|ghp_[A-Za-z0-9]{20,}|xox[baprs]-[A-Za-z0-9-]{20,})\b", re.IGNORECASE),
    re.compile(
        r"!\[[^\]\r\n]{0,2048}\]\(\s*https?:\/\/[^\s)]{0,2048}(?:token|secret|api[_-]?key|password|env)=", re.IGNORECASE
    ),
]

CODE_NOISE_PATTERNS: list[re.Pattern[str]] = [
    re.compile(r"^(show|print|repeat|echo)\s+(?:system\s+)?prompt$", re.IGNORECASE),
    re.compile(r"^(show|print|repeat|echo)\s+instructions?$", re.IGNORECASE),
]


def _classify_patterns(patterns_found: list[str]) -> EventType:
    """Heuristic mapping from matched pattern strings to event type."""
    joined = " ".join(patterns_found).lower()
    if any(
        w in joined
        for w in (
            "secret",
            "token",
            "api",
            "password",
            "credential",
            ".env",
            "private",
            "id_rsa",
            "id_ed25519",
            "kubeconfig",
        )
    ):
        if any(
            w in joined
            for w in (
                "http",
                "webhook",
                "external",
                "remote",
                "attacker",
                "requestbin",
                "pastebin",
                "upload",
                "exfiltrate",
            )
        ):
            return EventType.DATA_EXFILTRATION
        return EventType.CREDENTIAL_EXFIL
    if any(w in joined for w in ("ignore", "disregard", "override", "new instructions")):
        return EventType.SYSTEM_OVERRIDE
    if any(w in joined for w in ("dan", "pretend", "act as", "you are now", "unlimited")):
        return EventType.JAILBREAK_ROLEPLAY
    if any(w in joined for w in ("reveal", "show", "print", "repeat", "echo", "prompt", "instructions")):
        return EventType.PROMPT_LEAKAGE
    if any(w in joined for w in ("base64", "javascript", "data:text")):
        return EventType.CODE_EXEC_RISK
    if "im_start" in joined or "inst" in joined or "system:" in joined:
        return EventType.CONTEXT_CONFUSION
    return EventType.UNKNOWN


def _run_patterns(
    text: str,
    patterns: list[re.Pattern[str]],
    max_chars: int = 32_000,
) -> list[str]:
    found: list[str] = []
    chunk = text[:max_chars]
    for pat in patterns:
        m = pat.search(chunk)
        if m:
            found.append(m.group(0)[:80])
    return found


def _excerpt_around(text: str, patterns: list[re.Pattern[str]], max_chars: int = 32_000) -> str:
    chunk = text[:max_chars]
    for pat in patterns:
        m = pat.search(chunk)
        if m:
            start = max(0, m.start() - 50)
            end = min(len(chunk), m.end() + 50)
            return chunk[start:end].replace("\n", " ")
    return ""


def _is_code_like(text: str) -> bool:
    sample = text[:4000]
    if not sample.strip():
        return False
    lines = sample.splitlines()[:80]
    score = 0
    for raw in lines:
        line = raw.strip()
        if not line:
            continue
        if re.search(r"^(//|#|/\*|\*|--)\s*", line):
            score += 1
        if re.search(r"[{}()[\];]", line):
            score += 1
        if re.search(r"(^|\s)(func|class|def|const|let|var|return|if|for|while|import|export|package)\b", line):
            score += 1
        if re.search(r"[:=]-?=?|=>", line):
            score += 1
        if re.search(r"\b[A-Za-z_]\w*\s*\(", line):
            score += 1
    return score >= 4


def _filter_code_noise(matches: list[str]) -> list[str]:
    out: list[str] = []
    for m in matches:
        mt = m.strip()
        if any(p.search(mt) for p in CODE_NOISE_PATTERNS):
            continue
        out.append(m)
    return out


def scan_text(text: str, source: str = "unknown", max_scan_chars: int = 32_000) -> ScanResult:
    """Tier-1 core scan (user input, RAG, conversation history)."""
    if not text:
        return ScanResult(source=source)
    t0 = time.monotonic()
    found = _run_patterns(text, CORE_PATTERNS, max_scan_chars)
    excerpt = _excerpt_around(text, CORE_PATTERNS, max_scan_chars) if found else ""
    event_type = _classify_patterns(found) if found else EventType.UNKNOWN
    confidence = min(0.5 + 0.15 * len(found), 1.0) if found else 0.0
    return ScanResult(
        detected=bool(found),
        patterns_found=found,
        source=source,
        excerpt=excerpt,
        tier="core",
        confidence=confidence,
        event_type=event_type,
        scanner_name="regex_core",
        latency_ms=(time.monotonic() - t0) * 1000,
    )


def scan_web_content(text: str, source: str = "web", max_scan_chars: int = 32_000) -> ScanResult:
    """Tier-1 + Tier-2 scan with normalization and base64 probing."""
    if not text:
        return ScanResult(source=source, tier="web")
    t0 = time.monotonic()
    # Check for zero-width cluster evasion on raw text before stripping
    raw_web_found = _run_patterns(text[:max_scan_chars], WEB_PATTERNS, max_scan_chars)
    normalized = normalize_for_scan(text)
    core_found = _run_patterns(normalized, CORE_PATTERNS, max_scan_chars)
    web_found = _run_patterns(normalized, WEB_PATTERNS, max_scan_chars)
    # Merge raw findings (zero-width cluster hits only fire on raw text)
    seen = set(web_found)
    for f in raw_web_found:
        if f not in seen:
            web_found.append(f)
            seen.add(f)
    b64_found = detect_base64_payloads(text, CORE_PATTERNS, max_chars=max_scan_chars)
    all_found = core_found + web_found + b64_found
    if _is_code_like(text):
        all_found = _filter_code_noise(all_found)
    excerpt = _excerpt_around(normalized, CORE_PATTERNS + WEB_PATTERNS, max_scan_chars) if all_found else ""
    event_type = _classify_patterns(all_found) if all_found else EventType.UNKNOWN
    confidence = min(0.5 + 0.12 * len(all_found), 1.0) if all_found else 0.0
    return ScanResult(
        detected=bool(all_found),
        patterns_found=all_found,
        source=source,
        excerpt=excerpt,
        tier="web",
        confidence=confidence,
        event_type=event_type,
        scanner_name="regex_web",
        latency_ms=(time.monotonic() - t0) * 1000,
    )


def scan_model_output(text: str, source: str = "model_output") -> ScanResult:
    """Tier-3 output guardrail: detect compliance with injected instructions."""
    if not text:
        return ScanResult(source=source, tier="output")
    t0 = time.monotonic()
    found = _run_patterns(text, OUTPUT_PATTERNS, max_chars=16_000)
    excerpt = _excerpt_around(text, OUTPUT_PATTERNS, max_chars=16_000) if found else ""
    confidence = min(0.6 + 0.15 * len(found), 1.0) if found else 0.0
    return ScanResult(
        detected=bool(found),
        patterns_found=found,
        source=source,
        excerpt=excerpt,
        tier="output",
        confidence=confidence,
        event_type=EventType.PROMPT_LEAKAGE if found else EventType.UNKNOWN,
        scanner_name="regex_output",
        latency_ms=(time.monotonic() - t0) * 1000,
    )


def redact_patterns(text: str, include_web: bool = False) -> str:
    """Replace all matching spans with [REDACTED]."""
    patterns = list(CORE_PATTERNS)
    if include_web:
        patterns.extend(WEB_PATTERNS)
    for pat in patterns:
        text = pat.sub("[REDACTED]", text)
    return text


# ---------------------------------------------------------------------------
# Batch helpers (RAG, web results, message lists)
# ---------------------------------------------------------------------------


def scan_and_filter_texts(
    texts: list[str],
    source_prefix: str = "chunk",
    action: str = "reduce",
    web: bool = False,
) -> tuple[list[str], bool, list[dict[str, Any]]]:
    """Scan a list of text blocks; filter/redact per action (reduce|block|log)."""
    scan_fn = scan_web_content if web else scan_text
    output: list[str] = []
    any_detected = False
    details: list[dict[str, Any]] = []
    for i, t in enumerate(texts):
        if not t:
            output.append(t)
            continue
        r = scan_fn(t, source=f"{source_prefix}_{i}")
        if r.detected:
            any_detected = True
            details.append({"index": i, "patterns": r.patterns_found, "source": r.source})
            if action == "reduce":
                output.append(redact_patterns(t, include_web=web))
            elif action == "block":
                continue
            else:
                output.append(t)
        else:
            output.append(t)
    return output, any_detected, details


def scan_messages(
    messages: list[dict[str, Any]],
) -> tuple[list[dict[str, Any]], bool, list[ScanResult]]:
    """Scan all roles in an OpenAI-shaped message list (all roles untrusted).

    Returns (sanitized_messages, any_detected, scan_results).
    """
    any_detected = False
    results: list[ScanResult] = []
    out: list[dict[str, Any]] = []

    for msg in messages:
        role = msg.get("role", "")
        new_msg = dict(msg)

        content = new_msg.get("content")
        if isinstance(content, str) and content:
            r = scan_text(content, source=f"msg_{role}")
            if r.detected:
                any_detected = True
                results.append(r)
                new_msg["content"] = redact_patterns(content)

        if role == "assistant" and new_msg.get("tool_calls"):
            tcs, tc_detected, tc_results = _scan_tool_calls(new_msg["tool_calls"])
            if tc_detected:
                any_detected = True
                results.extend(tc_results)
                new_msg["tool_calls"] = tcs

        out.append(new_msg)

    return out, any_detected, results


def _scan_tool_calls(
    tool_calls: list[dict[str, Any]],
) -> tuple[list[dict[str, Any]], bool, list[ScanResult]]:
    any_detected = False
    results: list[ScanResult] = []
    out: list[dict[str, Any]] = []
    for tc in tool_calls:
        tc_copy = dict(tc)
        fn = tc_copy.get("function")
        if isinstance(fn, dict) and isinstance(fn.get("arguments"), str):
            r = scan_text(fn["arguments"], source="tool_call_args")
            if r.detected:
                any_detected = True
                results.append(r)
                fn = dict(fn)
                fn["arguments"] = redact_patterns(fn["arguments"])
                tc_copy["function"] = fn
        out.append(tc_copy)
    return out, any_detected, results
